README.md
# raml-mocker-server
Node module to run server mocking API described in RAML files



Objective
---
This plugin make use of [raml-mocker](https://github.com/RePoChO/raml-mocker) module. It run a [Express](http://expressjs.com/) server and configure endpoints and responce from RAML files.

How to use
---

Basic, this will create server on port 3030

```javascript
var api = require('raml-mocker-server');

var options = {
    path: 'test/raml' // path to folder with RAML files
};

var callback = function (app){
	// Express app could be used to configure more paths
    console.log('All RAML files parsed and API endpoints defined');
};

// returns created server
var server = api(options, callback);
```

* [test/testApp.js](test/testApp.js)

Add API to existing web server:

* [test/testAppExpress.js](test/testAppExpress.js)



Options
---

| Name         | Type          | Default value | Description                                                  |
| ------------ | ------------- | ------------- | -------------------------------------------------------------|
| port         | number        | 3030          | defines server port                                          |
| path         | string        | 'raml'        | path to raml folder, relative to the execution context       |
| prefix       | string, array | ''            | prefixing all API endpoints described in RAML                |
| debug        | boolean       |               | enable logging debug info to the console                     |
| watch        | boolean       |               | enable watching on RAML files                                |
| staticPath   | string        |               | defines path to the static folder                            |
| prioritizeBy | string        | 'schema'      | defines the priority of the endpoint response if both 'schema' and 'example' are defined |
| app          | object        |               | if server already exists you can pass express app, Express app, no need to specify *port* and *staticPath* |



#### [Changelog](CHANGELOG.md)



CHANGELOG.md
## Changelog

### v 0.1.5 (2015-12-21)
* Add Content-Type: application/json header for response

### v 0.1.4 (2015-12-01)
* Add prioritizeBy option

### v 0.1.2 (2015-10-26)
* Add CORS middleware

### v 0.1.1 (2015-02-05)
* return "app" in the callback

### v 0.1.0 (2015-01-22)
* update dependencies
* prefix could be an array
* update tests



test/testApp.js
/* globals require, __dirname, console */
'use strict';


var api = require('..');
var http = require('http');
var Q = require('Q');

var port = '5050';


function get (path){
	var dfd =  Q.defer();

	var req = http.get('http://localhost:' + port + path, function(res) {
		res.setEncoding('utf8');
		res.on('data', function (body) {
			console.log('Got response: ' + res.statusCode);
			console.log(res.headers['content-type']);
			console.log(body);
			dfd.resolve();
		});
	}).on('error', function(e){
		dfd.reject(new Error(e.message));
	});

	req.end();

	return dfd.promise;
}


function cb (app){

	console.log('APIs are ready to use');

	// Add custome route
	app.get('/custom', function(req,res){
		res.status(200).send('custom url');
	});

	Q.all([
		get('/custom'),
		get('/test/12345/example.json'),
		get('/api/test/12345/objectDef'),
		get('/example'),
		get('/faker'),
		get('/schema-first'),
		get('/example-first')
	]).then(function(){
		console.log('DONE');
		server.close();
	});
}


var options = {
	path: 'test/raml',
	debug: true,
	watch: true,
	port: port,
	prefix: ['', '/api'],
	prioritizeBy: 'example',
	staticPath: __dirname
};


var server = api(options, cb);



test/testAppExpress.js
/* globals require, console */
'use strict';


var api = require('..');
var express = require('express');
var app = express();

var cb = function(){
	console.log('callback, stop the server');
	watcher.close();
	server.close();
};

var options = {
	path: 'test/raml',
	debug: true,
	watch: true,
	app: app,
	prefix: '/api'
};

var watcher = api(options, cb);


console.log('Listening on port 3000');
var server = app.listen(3000);




index.js
/* globals require, module, console  */
'use strict';

var _ = require('lodash');
var express = require('express');
var cors = require('cors');
var chokidar = require('chokidar');
var ramlMocker = require('raml-mocker');

var options;
var requestsMap = {};
var app;

var defaults = {
	port: 3030,
	path: 'raml',
	prefix: '',
	prioritizeBy: 'schema', // example, todo:order
	// watch: true, // watching
	// debug: true, // shows logs
};

var colors = {default: '\x1b[0m', green: '\x1b[32m', qyan: '\x1b[36m'};

/**
 * Initializing RAML mocker server
 * @param  {Function} cb    callback executed af
 * @param  {Object}   prefs configuration object
 * @return {[type]}         [description]
 */
function init (prefs, callback) {
	options = _.extend(defaults, prefs);
	app = options.app || express();
	app.use(cors());

	ramlMocker.generate(options, process(function(requestsToMock){
		requestsToMock.forEach(function(reqToMock){
			addRoute(reqToMock);
		});
		log(new Array(30).join('='));
		log('%s[%s]%s API routes generated', colors.qyan, requestsToMock.length, colors.default);

		if(typeof callback === 'function'){
			callback(app);
		}
	}));

	var watcher = watch();
	var server = launchServer(app);

	function close () {
		if (watcher) { watcher.close(); }
		if (server) { server.close(); }
	}

	return {
		close: close,
		server: server,
		watcher: watcher
	};
}


function log (){
	if(options.debug){
		console.log.apply(console, arguments);
	}
}


function process (callback) {
	return function (requestsToMock) {
		requestsToMock.forEach(function(reqToMock){
			requestsMap[reqToMock.method + '!' + reqToMock.uri] = reqToMock;
		});
		if(typeof callback === 'function'){
			callback(requestsToMock);
		}
	};
}


function launchServer (app){

	// intercept OPTIONS method
	app.options('*', function(req, res) {
		res.status(200).send();
	});

	if (!options.app) {

		// set static location
		if(options.staticPath){
			app.use(express.static(options.staticPath));
		}


		// CORS middleware
		if(options.cors){
			app.use(function(req, res) {
				res.header('Access-Control-Allow-Origin', '*');
				res.header('Access-Control-Allow-Methods', 'GET,PUT,POST,DELETE');
				res.header('Access-Control-Allow-Headers', 'Content-Type, Authorization');
			});
		}


		console.log('Listening on port ' + options.port);
		return app.listen(options.port);
	}
}

function watch (){
	if(options.watch){
		var watcher = chokidar.watch(options.path, {ignored: /[\/\\]\./, persistent: true});
		watcher.on('change', function(path) {
			log('File modified: ', path);
			ramlMocker.generate(options, process());
		});
		return watcher;
	}
}

function addRoute (reqToMock){

	var prefixes = _.isArray(options.prefix) ? options.prefix : [options.prefix];
	var method = reqToMock.method;

	log(colors.qyan + method + colors.default, '\t', reqToMock.uri);

	prefixes.forEach(function(prefix){
		var uri = prefix + reqToMock.uri;

		app[method](uri, function(req,res){
			var mockObj = requestsMap[method + '!' + reqToMock.uri];
			if(mockObj){
				var mock = mockObj.mock && mockObj.mock();
				var example = mockObj.example && mockObj.example();
				var response =  options.prioritizeBy === 'example' ?
						example || mock :
						mock || example || '';

				res.setHeader('content-type', 'application/json');
				res.status(mockObj.defaultCode || 200).send(response);
			} else {
				res.status(404).send();
			}
			log(colors.green + method + colors.default, '\t', uri);
		});
	});
}


module.exports = init;
